### ISSUE TITLE
Split SQLStore into smaller per-domain services

### ISSUE THREAD

**Issue by zserge** [grafana/grafana#46522]
As the next step in making Grafana backend more modular - we attempt to split `sqlstore` packages into smaller ones.

Each package would serve one domain (i.e. "datasource" or "team" or "apikey").

Each package would expose a `Service` interface, with the implementation currently backed up by a SQL Store. Other store types might emerge in the future. Also, each package would provide a test double implementation of its Service.

# Domains

Here's a list of the possible domains and related SQL table names:

- [x] org (`org`) @idafurjes 
- [x] login (`login_attempt`) -- https://github.com/grafana/grafana/pull/54479
- [x] user (`user`, `org_user`) -- do we need a separate orguser service? @idafurjes https://github.com/grafana/grafana /pull/50502 #51912 https://github.com/grafana/grafana/pull/52442 https://github.com/grafana/grafana/pull/53226 https://github.com/grafana/grafana/pull/53204 https://github.com/grafana/grafana/pull/53298 https://github.com/grafana/grafana/pull/53595 https://github.com/grafana/grafana/pull/53717 https://github.com/grafana/grafana/pull/53726 https://github.com/grafana/grafana/pull/53734 https://github.com/grafana/grafana/pull/53788
- [x] tempuser (`temp_user`) -- also needs user for JOINs https://github.com/grafana/grafana/pull/53325
- [x] team (`team`, `team_member`, `team_role`)
- [X] pref (`preferences`) 🥧 https://github.com/grafana/grafana/pull/46843 https://github.com/grafana/grafana/pull/47870
- [X] star (`star`) 🥧  Ticket: https://github.com/grafana/grafana/issues/46538 PR: https://github.com/grafana/grafana/pull/45851
- [X]  dashboard (`dashboard`)
    *  https://github.com/grafana/grafana/pull/49347
    * https://github.com/grafana/grafana/pull/49175
    * https://github.com/grafana/grafana/pull/49170
    * https://github.com/grafana/grafana/pull/48971
    * https://github.com/grafana/grafana/pull/49962
    * https://github.com/grafana/grafana/pull/50091
- [X]  dashboard/snapshot (`dashboard_snapshot`) - @mildwonkey 
    * https://github.com/grafana/grafana/pull/50727 
- [X] dashboard/version (`dashboard_version`) 
    * https://github.com/grafana/grafana/pull/49138 
    * https://github.com/grafana/grafana/pull/49610 
    * https://github.com/grafana/grafana/pull/49967
    * https://github.com/grafana/grafana/pull/50287
- [x] datasource (`data_source`) - @mildwonkey 
- [x]  playlist (`playlist`, `playlist_item`) 🥧 https://github.com/grafana/grafana/pull/52252
- [x]  apikey (`apikey`) 🥧 @zserge - https://github.com/grafana/grafana/pull/52781

# Other services that should be refactored individually:

- [x] stat
- [x] quota (`quota`)

Here, 🥧 indicates that the domain is relatively easy and has no heavy external dependencies.

We should also create a minimal sub-interface from the SQL Store to limit its use to managing database connections/sessions and no business logic beyond that.

We might also want to move `wire.go` to the top level of `pkg/` to let it instantiate Service implementations from the `internal` subpackages.

For each domain we should announce the migration to everyone who might be impacted and:
1. Agree on an interface for the Service
2. Agree on the DTOs. These DTOs would be different from DTOs in the models package and would not contain `Result` field.
3. Copy (not move) the implementation of the Service/Store from the sqlstore package
4. Create (possibly - generate) a test double
5. At this point old implementation still works, but the new implementation can be used as well, since it uses the same database
6. Switch APIs and Services to use the new implementation
7. Migrate the tests (remove old ones, write new ones in a more reasonable manner)
8. Fix Enterprise to use the new implementation
9. Remove the old implementation from SQLStore 

Before starting a store/service for each domain it might make sense to include relates old sqlstore files into CODEOWNERS to get informed when someone changes them during the refactoring.

**Comment by sdboyer** [comment]
Great to see something written up about this, and happy to see the kinds of things you're thinking about!

Is a design doc coming?

> Each package would serve one domain (i.e. "datasource" or "team" or "apikey").

What's the motivation for the multilevel grouping? e.g., why group `team` and `team_member`, rather than having a flat space with them all as siblings?

> We might also want to move `wire.go` to the top level of `pkg/` to let it instantiate Service implementations from the `internal` subpackages.

In general i'm 🎉 🎉 on refactoring where wire sets go - clustering them in the http server is awkward on a good day. When i initially thought about it a month ago, i also imagined a way of pulling things out of `internal` dirs.

But, when i thought about it more, i came to the conclusion that it can't work - since wire is still all just generating normal Go code, code in e.g. `pkg/wire.go` won't be able to call e.g. `pkg/services/star/internal/store.ProvideService`. Did i miss something? Or is that not the kind of pattern you mean by "`internal` subpackages"?

> Copy (not move) the implementation of the Service/Store from the sqlstore package

Why not a type alias? This kind of refactoring was a major motivation for introducing them originally

**Comment by zserge** [comment]
A design doc is being written, and some rough PoC have been done.

`team` and `team_member` are two SQL tables, that team "service" currently deals with. The reason there are multiple tables for some domains is normalisation. These might be further split into separate domains, but for now if there are INNER JOINs everywhere - I'd keep it as. I'd prefer to avoid large _and_ smart refactorings.

As for wire - ah, snap! Yeah, it might be possible to generate bits of wire into pkg/service/star etc but probably not worth it. Yet, I'd also like to see implementations inside the internal packages of service, if there are Go mechanisms for that.

As for "copy, not move" there should be some clarification. When we refer to "Store" we mean the dumbest possible, most likely - private - interface that resembles CRUD. This "Store" would be executing SQL statements in its default implementation. I can easily see other "store" backends introduced later.

When we say "Service" we mean a public API that others can inject and use. This would include all the business logic that we currently have torn between the API and the SQL Store.

So, to fully unwind the ball of `sqlstore` we plan to create independent per-domain packages (not necessarily "atomic" or CRUD-like, it might be a "user" domain which deals with users, orgs and all other related entities). The smaller they are - the better, but rather than making 5 small packages that all depend on each other and can't be used individually - I'd better leave them together in one package.

Then, in each per-domain package we define a Service API that matches the funcionality of API/sqlstore logic (i.e. current handling of various search options, or icky "merging" of the perferences, or dealing with "temporary" users). It's the functionality that we need to keep, but Store is definitely the wrong place for it. Thus - Service does the dirty job and Store merely performs CRUD.

Treating things like this leaves us with "copy" strategy, where we take API+sqlstore code and split it into "store" code and "service" code. Aliasing does not help much here.

**Comment by sdboyer** [comment]
> domains

I'm seeing this word a lot, but no clear definition for it. (I'm inferring that usage has developed during private discussions). If it's going to be used as a basis for categorization i think a clear definition is important to have - one that can be used to unambiguously decide why a given object is in one domain vs. another.

> some rough PoC have been done.

Are they merged in `main` already?

> As for wire - ah, snap! Yeah, it might be possible to generate bits of wire into pkg/service/star etc but probably not worth it. Yet, I'd also like to see implementations inside the internal packages of service, if there are Go mechanisms for that.

The only way i've been able to think of doing it is going for a radically different package structure where _most_ of Grafana is under a top-level `internal` dir in the repo, and an artificially-constructed public surface area under `pkg/` consisting mostly of type aliases. But that approach really can't accommodate a sane, module-based dependency by enterprise - at least, that i was able to come up with.

The only other thing i was able to think of is the hilarious/horrible antipattern that is "the org private repo" - `github.com/grafana/internal` - which would allow us to have code that's internal to both grafana/grafana and grafana/grafana-enterprise. But that adds another repo to the mix, which creates far more problems than it solves.

So i ultimately dismissed direct use of `internal/` packages as an unreachable dream, and accepted that things would generally have to be public under `pkg/`. After reaching that conclusion, the next possibility that began to emerge was using compgen's `not` rules (not demonstrated in the hackathon video, but [here's one](https://github.com/grafana/hackathon-2022-03-compgen/blob/main/dirfixtures/basic/compgen.yaml#L11-L14) - "the component must not contain a reference to the given symbol") to start limiting interactions with things i'd ideally like to be `internal/`. I'm not sold myself - it's not nearly as clean as a simple `internal/` rule, and doesn't have the same meaning for the public looking at Grafana's repo, but it does at least seem to be in the right ballpark.

> When we refer to "Store" we mean the dumbest possible, most likely - private - interface that resembles CRUD. This "Store" would be executing SQL statements in its default implementation. I can easily see other "store" backends introduced later.

Not quite clear on how a Store interface could be simultaneously be a) private, b) non-`internal` (per above), c) within a domain, and d) support having multiple backends in a way that i imagine would have to have some uniform controls across the domains (e.g., a central control point for deciding between backend X and Y for all domains)

> `team` and `team_member` are two SQL tables, that team "service" currently deals with. The reason there are multiple tables for some domains is normalisation. These might be further split into separate domains, but for now if there are INNER JOINs everywhere - I'd keep it as. I'd prefer to avoid large and smart refactorings.

While i think i understand the general sentiment, and i certainly agree that changing fewer things at once is preferable, i can't say i really follow this reasoning. An additional grouping concept like "domain" on top of the logical objects themselves is nontrivial, and needs a clear reason to exist, even if we're hoping it only exists temporarily. (And it's not clear whether domain is  the case.)

* I'm not sure what 'team "service"' refers to - do you mean [this interface](https://github.com/grafana/grafana/blob/e135b8531a7e0e9606f58bfd4b291cec1394c3d5/pkg/services/sqlstore/team.go#L29-L39)?
* Are you proposing a firm rule that disallows JOINs across tables in different domains?
* Degree of [de]normalization in a SQL db layout seems like a somewhat artificial basis for choosing, because i think that's generally driven by a natural tendency to avoid duplicating data (e.g., there shouldn't be a row in `team` for each `team_role` because in a SQL table, we'd have to repeat the fields multiple times). But those relational constraints are limited to SQL representation, and are pretty orthogonal to our object model.

> The smaller they are - the better, but rather than making 5 small packages that all depend on each other and can't be used individually - I'd better leave them together in one package.

In what sense "can't" we use the objects individually? A hard constraint, like import cycles? An example would be helpful

**Comment by sdboyer** [comment]
> Each package would serve one domain (i.e. "datasource" or "team" or "apikey").

> Each package would expose a Sevice interface, with the implementation currently backed up by a SQL Store. 

Actually...rereading all this, if we're saying there's just one Service per domain here, and that all objects in a domain could plausibly be combined together - denormalized, if you prefer - into a single overall object - e.g., a `Team` struct that includes a slice of team roles and a slice of team members, i think that addresses my main concern about "domain" introducing an arbitrary layer of grouping. (Though, still not sold on it as the term to use - this discussion being a case in point, as it suggested a broader, loosely-defined grouping than you seemed to intend.)

If the rule is "all parts of domain must be representable as a single object", does that suggest `org_user` should be its own domain?

Still looking forward to the design doc :)

**Comment by sdboyer** [comment]
Haven't gotten a response on [some](https://docs.google.com/document/d/1t9sgMGlHxwU0SBvmdkVX3nchz81ME7PEGihyg0NEzxI/edit?disco=AAAAXBcgza4) [comments](https://docs.google.com/document/d/1t9sgMGlHxwU0SBvmdkVX3nchz81ME7PEGihyg0NEzxI/edit?disco=AAAAXBcgzZ0) in the design doc that i think need to be addressed before starting to merge PRs, so reiterating them here:

### "Domain"

"Domain" is not the best term. I realize, naming, bikeshedding, but for the amount of work that is going into this, i don't buy that these terms will just be around temporarily. "Entity" or "entity type" seems considerably better.

### Enforcement

I generally like these conventions, but not their lack of enforcement. The proposal needs to include some actual enforcement mechanism.

In a codebase our size, half-followed conventions end up just adding more incidental complexity. especially when the conventions are decently complex and interlinked, like these. Things like Bazel have lots of capabilities around doing this, specifically because they keep larger codebases maintainable.

I'd like it to be compgen - it will simplify future conversations about merging different component models. But it doesn't have to be compgen, and compgen won't be sufficient to express everything relevant here, particularly not with its current ruleset. But, something that defines and enforces these conventions, independent of any particular instance.
